onnx/large_helper.py

```diff
@@ -0,0 +1,333 @@
+# Copyright (c) ONNX Project Contributors
+#
+# SPDX-License-Identifier: Apache-2.0
+
+import enum
+import os
+from typing import Any, Dict, Iterable, Optional, Set, Tuple
+
+import numpy as np
+
+from onnx import (
+    GraphProto,
+    ModelProto,
+    StringStringEntryProto,
+    TensorProto,
+    checker,
+    load_model,
+)
+from onnx.external_data_helper import (
+    ExternalDataInfo,
+    _get_all_tensors,
+    _sanitize_path,
+    uses_external_data,
+)
+from onnx.helper import make_model
+
+
+def make_large_tensor_proto(
+    location: str, name: str, tensor_type: int, shape: Tuple[int, ...]
+) -> TensorProto:
+    """
+    Create an external tensor.
+
+    :param location: unique identifier (not necessary a path)
+    :param name: initializer name in the graph
+    :param tensor_type: onnx type
+    :param shape: shape the of the initializer
+    :return: the created tensor
+    """
+    tensor_location = location
+    tensor = TensorProto()
+    tensor.name = name
+
+    del tensor.external_data[:]
+    tensor.data_location = TensorProto.EXTERNAL
+    for k, v in {
+        "location": tensor_location,
+        "offset": None,
+        "length": None,
+        "checksum": None,
+        "basepath": None,
+    }.items():
+        if v is not None:
+            entry = tensor.external_data.add()
+            entry.key = k
+            entry.value = str(v)
+
+    tensor.data_type = tensor_type
+    tensor.dims.extend(shape)
+    return tensor
+
+
+class LargeModelFileFormat(enum.IntEnum):
+    # One file for all the weights.
+    SINGLE_TENSOR_FILE = 2
+
+    # Multiple files, one file with extension `.onnx` for the
+    # main graph and one file `.weight` for every large initializer.
+    # It uses the same format as `write_external_data_tensors`.
+    ONE_TENSOR_PER_FILE = 3
+
+
+class LargeModelContainer:
+    """
+    Implements an API to save large onnx models.
+    Avoids copying large initializers when defining the model.
+    """
+
+    def __init__(self):
+        self.model_proto_: Optional[ModelProto] = None
+        self.large_initializers: Dict[str, np.ndarray] = {}
+
+    def check_model(self):
+        if self.model_proto is not None:
+            checker.check_model(self.model_proto)
+
+    @property
+    def model_proto(self) -> ModelProto:
+        if self.model_proto_ is None:
+            raise RuntimeError("LargeModelContainer is empty.")
+        return self.model_proto_
+
+    @model_proto.setter
+    def model_proto(self, model_proto: ModelProto):
+        self.model_proto_ = model_proto
+        self.graphs_ = list(self.enumerate_graph_protos())
+
+    @staticmethod
+    def _enumerate_subgraphs(graph):
```

Comment:
I would avoid static methods. They can be private functions in the module

Reply:
done